### PR TITLE
Boosting refactoring

### PR DESCRIPTION
**Pull request by Onix22888**
# What this PR does?
Gradient boosting refactoring

## Issue reference
#4 

## Explanation
Remove class BaseEstimator

# How to get?
`git clone --recursive -b boosting_refactoring git@github.com:Bellator95/gradient_boosting.git`

### CODE REVIEW

File: gradient_boosting.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import numpy as np
+
+from sklearn.base import BaseEstimator
+from sklearn.base import RegressorMixin
+from tree import DecisionTreeRegressor
+
+
+class BaseGradientBoosting(BaseEstimator):
```
Comment (shaxov) [inline gradient_boosting.py]: Why did you not replace this class with yours?

File: gradient_boosting.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            y_pred += self.learning_rate[i] * yi_pred
+
+    def predict(self, X):
+        y_pred = np.zeros(shape=(X.shape[0], 1)) + self._init_pred
+        for i, est in enumerate(self._estimators):
+            y_pred += self.learning_rate[i] * est.predict(X)
+        return y_pred
+
+
+class GradientBoostingRegressor(BaseGradientBoosting, RegressorMixin):
```
Comment (shaxov) [inline gradient_boosting.py]: The same as previous. To refactor means "to provide your changes in current code" instead of "to create another code below".

File: gradient_boosting.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                 learning_rate=None):
+
+        super().__init__(
+            base_estimator=base_estimator,
+            n_estimators=n_estimators,
+            learning_rate=learning_rate
+        )
+
+
+if __name__ == '__main__':
```
Comment (shaxov) [inline gradient_boosting.py:46]: To read about the usage of this construction.
https://stackoverflow.com/questions/419163/what-does-if-name-main-do

File: gradient_boosting.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    import numpy as np
+
+    from sklearn.base import RegressorMixin
+    from tree import DecisionTreeRegressor
+
+
+    class BaseGradientBoostingRegressor:
+
+        def __init__(self,
```
Comment (shaxov) [inline gradient_boosting.py]: To add the `max_depth` parameter into the constructor interface instead of fixing it as `mx_depth=2` inside.

File: gradient_boosting.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                     n_estimators: int,
+                     learning_rate=None):
+
+            self.estimator = DecisionTreeRegressor(max_depth=2)
+            self.n_estimators = n_estimators
+            self.learning_rate = [learning_rate] * n_estimators
+
+            self._base_class = self.estimator.__class__
```
Comment (shaxov) [inline gradient_boosting.py]: You know the estimator's class! It is `DecisionTreeRegressor` which you defined above.

File: gradient_boosting.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    from tree import DecisionTreeRegressor
+
+
+    class BaseGradientBoostingRegressor:
+
+        def __init__(self,
+                     n_estimators: int,
+                     learning_rate=None):
+
+            self.estimator = DecisionTreeRegressor(max_depth=2)
```
Comment (shaxov) [inline gradient_boosting.py]: You need to define the estimator class instead of the object. You need a class  (not object) to define many objects in `_set_estimator()` method.

File: gradient_boosting.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    y_pred = gb.predict(x_cont)
+    y_pred2 = gb2.predict(x_cont)
+
+    fig, ax = plt.subplots(1, 2, figsize=(10, 5), sharey=True)
+    ax[0].plot(x, y, 'o')
+    ax[0].plot(x_cont, y_pred, 'r')
+    ax[0].plot(x_cont, y_pred2, 'orange')
+
+    ax[1].plot(x, y.ravel() - gb.predict(x).ravel(), 'o')
+    plt.show()
```
Comment (shaxov) [inline gradient_boosting.py]: To add a new line after the last row (PEP8).

File: gradient_boosting.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+    from sklearn.base import RegressorMixin
+    from tree import DecisionTreeRegressor
+
+
+    class BaseGradientBoostingRegressor:
+
+        def __init__(self,
+                     n_estimators: int,
+                     learning_rate=None):
```
Comment (shaxov) [inline gradient_boosting.py]: Will this code work if `learning_rate` default value is not changed?

File: gradient_boosting.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    import numpy as np
+
+    from sklearn.base import RegressorMixin
+    from tree import DecisionTreeRegressor
+
+
+    class BaseGradientBoostingRegressor:
+
+        def __init__(self,
+                     n_estimators: int,
```
Comment (shaxov) [inline gradient_boosting.py]: To provide the default value for parameter `n_estimators`.

File: gradient_boosting.py [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+            self.estimator = DecisionTreeRegressor(max_depth=2)
+            self.n_estimators = n_estimators
+            self.learning_rate = [learning_rate] * n_estimators
+
+            self._base_class = self.estimator.__class__
+            self._estimator_params = self.estimator.get_params()
+
+            self._set_estimator()
+
+        def _set_estimator(self):
```
Comment (shaxov) [inline gradient_boosting.py]: We set many estimators instead of one estimator. To rename this method.

File: gradient_boosting.py [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+            super().__init__(
+                n_estimators=n_estimators,
+                learning_rate=learning_rate
+            )
+
+
+if __name__ == '__main__':
+    from sklearn.utils import shuffle
+    import matplotlib.pyplot as plt
+    from sklearn.ensemble import GradientBoostingRegressor as GradientBoostingRegressor2
```
Comment (shaxov) [inline gradient_boosting.py]: To use the appropriate name for the baseline estimator. For instance `BaselineRegressor`.

File: gradient_boosting.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+
+    # from sklearn.tree import DecisionTreeRegressor
+
+    params = {'n_estimators': 5, 'max_depth': 2,
+              'learning_rate': 1., 'loss': 'ls'}
+    gb2 = GradientBoostingRegressor2(**params)
+
+    gb = GradientBoostingRegressor(5, learning_rate=1.)
```
Comment (shaxov) [inline gradient_boosting.py]: To make sure that the parameters of your estimator and baseline estimator are the same. Make it more clear in code.

**Review by shaxov — CHANGES_REQUESTED** [review]

File: gradient_boosting.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
+import numpy as np
+
+from sklearn.base import RegressorMixin
+from tree import DecisionTreeRegressor
+
+
+class BaseGradientBoostingRegressor:
+
+    def __init__(self,
+                 n_estimators: 5,
```
Comment (shaxov) [inline gradient_boosting.py]: **This is annotation, but not the default parameter**. Default parameter can be set using '=' (for example `n_estimators=100`. Please, read about [annotation](https://www.python.org/dev/peps/pep-3107/) in python and module [typing](https://docs.python.org/3/library/typing.html).

File: gradient_boosting.py [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+                 max_depth: int,
```
Comment (shaxov) [inline gradient_boosting.py]: Please, set `None` as a default parameter for `max_depth`, which means no constraints on depth.

File: gradient_boosting.py [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+                 learning_rate=None):
```
Comment (shaxov) [inline gradient_boosting.py]: Does it work if you do not specify `learning_rate` parameter? I mean, that if `learning_rate` is `None`, will the code be runnable?

File: gradient_boosting.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+        self.estimator = DecisionTreeRegressor
```
Comment (shaxov) [inline gradient_boosting.py]: You don't use variable `estimator` outside the class. Please, make it protected via a single underscore (`self._estimator`).

File: gradient_boosting.py [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.learning_rate = [learning_rate] * n_estimators
+
+        self._estimator_params = self.estimator().get_params()
```
Comment (shaxov) [inline gradient_boosting.py]: Here, you get default parameters of `DecisionTreeRegressor` instead of taking the parameters which are passed via the constructor.

File: gradient_boosting.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+        self._set_estimators()
+
+    def _set_estimators(self):
+        est_params = self._estimator_params
+
+        self._estimators = [self.estimator(**est_params)
+                            for _ in range(self.n_estimators)]
+
+    def fit(self, X, y):
+
```
Comment (shaxov) [inline gradient_boosting.py]: Please, remove the redundant line.

File: gradient_boosting.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+    def __init__(self,
+                 n_estimators: 5,
+                 max_depth: int,
+                 learning_rate=None):
+
+        self.estimator = DecisionTreeRegressor
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.learning_rate = [learning_rate] * n_estimators
```
Comment (shaxov) [inline gradient_boosting.py]: This behavior is not obvious. The user specifies `learning_rate` as `float` but inside it transforms into the `list()`. Please, create separate protected variable `_learning_rates_array` and use it for storing list of learning rates. But still, keep the variable `learning_rate` with the specified float value which is taken from the constructor.

File: gradient_boosting.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+        self._estimator_params = self.estimator().get_params()
+
+        self._set_estimators()
+
+    def _set_estimators(self):
+        est_params = self._estimator_params
+
```
Comment (shaxov) [inline gradient_boosting.py]: Remove redundant line.

File: gradient_boosting.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+            y_pred += self.learning_rate[i] * yi_pred
+
+    def predict(self, X):
+        y_pred = np.zeros(shape=(X.shape[0], 1)) + self._init_pred
+        for i, est in enumerate(self._estimators):
+            y_pred += self.learning_rate[i] * est.predict(X)
+        return y_pred
+
+
+class GradientBoostingRegressor(BaseGradientBoostingRegressor, RegressorMixin):
```
Comment (shaxov) [inline gradient_boosting.py]: Note, that this class contains no specific logic. It just adds the interface of regressor mixin. Please, remove it, and use just base class (without word 'Base' in the name) also with regressor mixin interface.

**Review by shaxov — CHANGES_REQUESTED** [review]

File: gradient_boosting.py (hunk 15 above)
Comment (Onix22888) [reply gradient_boosting.py]: No, it will not run. So I will set a default value.

File: gradient_boosting.py (hunk 17 above)
Comment (Onix22888) [reply gradient_boosting.py]: Yes, but why is it wrong? We are working with that estimator and here we get his parameters.

Comment (shaxov) [reply gradient_boosting.py]: Tell me, where parameter `n_estimators` is used? Except for line 17. The same for parameter `max_depth`.

**Review by shaxov — APPROVED** [review]
